#!/home/ueb/repo/peeva/eva_params.R
# revision $Revision-Id$ - $Date$ by $Commiter$
#
# SCRIPT: eva_params.R
# SCOPE: to be called from eva_main.R
# Author: Xavier de Pedro (2011-2013)
#         xavier.depedro@vhir.org
#         http://ueb.vhir.org/tools/EVA
###################################################

## Dual Licensed under the following copyright licenses:
## a) Creative Commons Attribution-ShareAlike 3.0 License : cc-by-sa 3.0
## b) GNU/LGPL : http://www.gnu.org/copyleft/lesser.html
##
##
##
## Acknowledgements: 
## * Michael Zeller, Harry Mangalam and Vinh Nguyen for http://blog.nguyenvq.com
## * Alex Sánchez Pla, for comments and feedback
## * Aleix Ruiz de Villa, for comments and feedback
## * Josep Lluis Mosquera, for comments and feedback
## * Ferran Briansó Castilla, for comments and feedback

##############################################################

# Roadmap (Accepted ToDo things):
# ----------------------
# (See also http://ueb.vhir.org/PendentsEVA )
##
## * [FIXED] investigate issue with pcr duplicates not being removed by samtools. See it here, for instance, for the same testset:
##    load in igv the test bam files and search for for this snp in brca1: chr17:41,196,307-41,197,202
##
## * April 16th, 2013. Variant Filtering bug:
##    st.vf.Q (e.g.=10) seems to do nothing filtering out variants with less than this base quality in samtools varFilter. 
##    Therefore filter added to the Variant Calling step instead (samtools mpileup also accepts this command)
##
## * [DONE for target genes] rework the csv files generated by annovar so that indel is on its own column, and DP=, and all the other INFO
##    use one column each, so that the researcher can filter out more the list of variants, for fileds like DP (Read Depth, etc). 
##
## * [DONE] move count reads function to the end of the workflow
##
## * [DONE] Improve count reads so that it takes less time (and/or memmory to finish post-processing data with R)
## ** Avoid errors at countreads:
##  #) at "b4. Re-order results in R":
## -----
##  # [SOLVED] as.data.frame.table or as.data.frame.matrix !!!!!!
##  f2p.cr.dft <- as.data.frame.table(f2p.cr.table)
##  # We need to re-sort in a similar way to what the ftable was
##  f2p.cr.dft <- f2p.cr.dft[order(f2p.cr.dft$GeneSymbol, f2p.cr.dft$unitN),]
##  # And last, we remove all row with 0 in the column Counts
##  f2p.cr.dft <- f2p.cr.dft[!f2p.cr.dft$Freq==0,]
## ------
## maybe solve it by using ddplyr or aggregate functions?
## See https://ueb.vhir.org/Chunks+de+codi+interessants&structure=UEB&page_ref_id=90#Aggregate_with_R 
## -----
##
## * Performance Improvements Pending:
##  #) [DONE] split the bam files by chr, so that variant calling with samtools can be run in parallel for one patient (each chr to a differnt cpu, etc)
##      However, the result is not clear, since there was some error in the MQ of one variant, so I can't trust blidnly this approach of using xargs, etc.
##
## * Count reads again from the results for the target genes with variants found, 
##    so that number of reads count can be split between 'Wild type' and 'Variant' (and therefore, '% or variant' can be computed).
##    as requested, like it's shown in Walsh et al. ( http://www.pnas.org/cgi/doi/10.1073/pnas.1007983107 )
##  a) Check the option to do it with vcftools: http://vcftools.sourceforge.net/options.html#stats 
##
## * For the minorallele info requested by some researchers, investigate "Ensembl Variation 71" (Releases in April 2013)
##    mart <- useMart(biomart="ENSEMBL_MART_ENSEMBL", host="www.ensembl.org", path="/biomart/martservice")
##    "Minor allele", "1000 genomes global MAF" and "1000 genomes global MAC" have been renamed to "Minor allele (ALL)", "1000 genomes global MAF (ALL)" and "1000 genomes global MAC (ALL)".
##    See http://www.ensembl.org/info/website/news.html
##
##
## * add dbNSFP annotation to SnpEff
## * add GWAS Catalog annotation
## * implement first steps with GATK successfully
## * assess adding a new step for local realignment with InGap
##
## <--- End of Roadmap

# 0. Basic startup params
# ---------------------------------------------------------------------
startdate <- paste(format(Sys.Date(), "%y%m%d"), sep="")

p_test     = 2 # 0-1-2; ### Is this a test run? ###
                # 0 = normal run
                # 1 = test run, so that use the predefined values for a test run, with samples: ; 
                # 2 = test run, so that use the predefined values for a test run; 
if (p_test==1) {
  path_input_absolute <- "0" # Define whether the p_input is absolute or relative
  p_input    <- "test_in1" #"/mnt/magatzem02/tmp/run_sara_293a/dir_in_293a3" #"test_in2" # "../test_in2"  # "test_in"
  p_in.suffix <- "_sequence" #"_sequence" # "" # This is the suffix of all input filenames (without extension) used for the pipeline to process
  p_in.ext    <- ".fastq" #".fastq" # ".fa" ".sam" ".bam" # This is the .extension of all files used as input for the pipeline to process
  p_output   <- "test_out1" #"/mnt/magatzem02/tmp/run_sara_293a/dir_out_293a3b" #"test_out2" # "../test_out2" # "test_out"
  p_f_my_rs  <- "file_my_rs.txt" # In p_input. Needed by SnpEff to filter for the target genes before the report (well, filter for the potential snp rs codes in those genes)
  p_label    <- "test1_06" #"testrunGATK1" # "testsnpEffCountReads_a" "test-121002" # "test-foo"        # Run Label for output filenames
  p_desc     <- "Testing the overall pipeline starting from scratch, with 
#                                sample_a_1_sequence.fastq 
#                                sample_a_2_sequence.fastq
#                                sample_b_1_sequence.fastq 
#                                sample_b_2_sequence.fastq
                     test1_06: As testset05 but playing with bed files to filter for intersecting intervals from target genes" 
#                      testset05: As testset4 but going back to test the variant filtering process with samtools, since we get zero results with current default settings
#                               varFilter -Q10 -d2  -a1 -D10000000 -S1000 ->   72 records in filtered.vcf
#                               varFilter -Q5  -d1  -a2 -D10000000 -S1000 ->   32 records in filtered.vcf
#                               varFilter -Q5  -d1  -a1 -D10000000 -S1000 ->  716 records in filtered.vcf
#                               varFilter -Q5  -d2  -a1 -D10000000 -S1000 ->   74 records in filtered.vcf
#                               varFilter -Q10 -d2  -a1 -D10000000 -S1000 ->   74 records in filtered.vcf
#                               varFilter -Q10 -d2  -a2 -D10000000 -S1000 ->   32 records in filtered.vcf
#                               varFilter -Q1  -d2  -a2 -D10000000 -S1    ->   32 records in filtered.vcf
#                               varFilter -Q10 -d2  -a2 -D10000000 -S1    ->   32 records in filtered.vcf
#                               varFilter -Q10 -d10 -a1 -D10000000 -S1000 ->    0 records in filtered.vcf
#                               varFilter -Q10 -d5  -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
#                               varFilter -Q1  -d10 -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
#                               varFilter -Q5  -d10 -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
#                               varFilter -Q10 -d10 -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
#                               "
  #                   testset04: As testset3 but playing with bed files to filter for intersecting intervals from target genes" # "Testing the new testset1 from the source merged sam file. See http://ueb.vhir.org/PendentsEVA and http://ueb.vhir.org/sheet3" #"Testing gatk files etc"# Testing The issue in snpEff Count Reads with first set: s_7_m11_149b_merged12_00-01M.sam"
  #                 testset03: As testset2 but variant calling: bcftools without -A, to see if X go away from alternative nucleotides in vcf files" # "Testing the new testset1 from the source merged sam file. See http://ueb.vhir.org/PendentsEVA and http://ueb.vhir.org/sheet3" #"Testing gatk files etc"# Testing The issue in snpEff Count Reads with first set: s_7_m11_149b_merged12_00-01M.sam"
  #    "testset99: As in testset99-1, but with new version of annovar and associated db:
  #             * Annovar from 2013 02 21 (instead of 2012 10).
  #             * 131M Feb 11  2013 hg19_refGeneMrna.fa
  #             *  54M Mar  7  2013 hg19_cosmic63.txt
  #             * 6.2M Mar  7  2013 hg19_cosmic63.txt.idx
  #             *  17K Feb 26  2013 hg19_MT_GRCh37_ensGeneMrna.fa
  #             * 3.2K Feb 26  2013 hg19_MT_GRCh37_ensGene.txt
  #             * 131M Feb 11  2013 hg19_refGeneMrna.fa
  #             *  12M Mar 18  2013 hg19_refGene.txt
  #             *  29M Mar 18  2013 hg19_refLink.txt
  #             * 845M May 25  2012 hg19_snp132NonFlagged.txt.gz
  #             * 846M Oct  1  2011 hg19_snp132.txt.gz
  #             * 1.3G May 25  2012 hg19_snp135NonFlagged.txt.gz
  #             * 1.3G Feb 23  2012 hg19_snp135.txt.gz
  #             * 1.4G Jan  9  2013 hg19_snp137NonFlagged.txt.gz
  #             * 1.4G Jan  7  2013 hg19_snp137.txt.gz
  #             "
  
  #                 testset2: As testset1 but with my_rs.txt added to filter for known snps from target genes" # "Testing the new testset1 from the source merged sam file. See http://ueb.vhir.org/PendentsEVA and http://ueb.vhir.org/sheet3" #"Testing gatk files etc"# Testing The issue in snpEff Count Reads with first set: s_7_m11_149b_merged12_00-01M.sam"
  p_keep     <- TRUE # Enable if run through editor and you want to keep temp files
  p_showc    <- TRUE # Enable if you want to see the commands literally that are run in the command line
  p_st.vf.Q  <-      10 # Samtools vcftools.pl varFilter Q param: Minimum base quality for a base to be considered: 10 (same for small testing sets). 10 (90% accuracy)|20 (99.9%)|30 (99.99%)|40 (99.999%)|50 (99.9999%). See Wikipedia.".
  p_st.vf.d  <-       1 #(2) Samtools vcftools.pl varFilter d param: Minimum read depth (coverage) to call a SNP: 10-15 (1-2 for small testing sets)".
  p_st.vf.a  <-       1 #(1) Samtools vcftools.pl varFilter a param: Minimum number of alternate bases: 2-3 (1-2 for small testing sets)".
  p_st.vf.D  <-10000000 # Samtools vcftools.pl varFilter D param: Maximum read depth (coverage) to call a SNP: 10000000 (same for small testing sets)".
  p_st.vf.S  <-    1000 # Samtools vcftools.pl varFilter S param: minimum SNP quality: 1000 (same for small testing sets). The smaller, the better (more precise, more quality in the SNP). High values are too permissive."
  #p_filter   <- ""   
  p_filter   <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1:\\|^BRCA1\t\\|^BRCA2\t\\|^CHEK2\t\\|^PALB2\t\\|^BRIP1\t\\|^TP53\t\\|^PTEN\t\\|^STK11\t\\|^CDH1\t\\|^ATM\t\\|^BARD1\t\\|^APC\t\\|^MLH1\t\\|^MRE11A\t\\|^MSH2\t\\|^MSH6\t\\|^MUTYH\t\\|^NBN\t\\|^PMS1\t\\|^PMS2\t\\|^RAD50\t\\|^RAD51D\t\\|^RAD51C\t\\|^XRCC2\t\\|^UIMC1\t\\|^FAM175A\t\\|^ERCC4\t\\|^RAD51\t\\|^RAD51B\t\\|^XRCC3\t\\|^FANCA\t\\|^FANCB\t\\|^FANCC\t\\|^FANCD2\t\\|^FANCE\t\\|^FANCF\t\\|^FANCG\t\\|^FANCI\t\\|^FANCL\t\\|^FANCM\t\\|^SLX4\t\\|^CASP8\t\\|^FGFR2\t\\|^TOX3\t\\|^MAP3K1\t\\|^MRPS30\t\\|^SLC4A7\t\\|^NEK10\t\\|^COX11\t\\|^ESR1\t\\|^CDKN2A\t\\|^CDKN2B\t\\|^ANKRD16\t\\|^FBXO18\t\\|^ZNF365\t\\|^ZMIZ1\t\\|^BABAM1\t\\|^LSP1\t\\|^ANKLE1\t\\|^TOPBP1\t\\|^BCCIP\t\\|^TP53BP1\t\\|\"BRCA1\"\\|\"BRCA2\"\\|\"CHEK2\"\\|\"PALB2\"\\|\"BRIP1\"\\|\"TP53\"\\|\"PTEN\"\\|\"STK11\"\\|\"CDH1\"\\|\"ATM\"\\|\"BARD1\"\\|\"APC\"\\|\"MLH1\"\\|\"MRE11A\"\\|\"MSH2\"\\|\"MSH6\"\\|\"MUTYH\"\\|\"NBN\"\\|\"PMS1\"\\|\"PMS2\"\\|\"RAD50\"\\|\"RAD51D\"\\|\"RAD51C\"\\|\"XRCC2\"\\|\"UIMC1\"\\|\"FAM175A\"\\|\"ERCC4\"\\|\"RAD51\"\\|\"RAD51B\"\\|\"XRCC3\"\\|\"FANCA\"\\|\"FANCB\"\\|\"FANCC\"\\|\"FANCD2\"\\|\"FANCE\"\\|\"FANCF\"\\|\"FANCG\"\\|\"FANCI\"\\|\"FANCL\"\\|\"FANCM\"\\|\"SLX4\"\\|\"CASP8\"\\|\"FGFR2\"\\|\"TOX3\"\\|\"MAP3K1\"\\|\"MRPS30\"\\|\"SLC4A7\"\\|\"NEK10\"\\|\"COX11\"\\|\"ESR1\"\\|\"CDKN2A\"\\|\"CDKN2B\"\\|\"ANKRD16\"\\|\"FBXO18\"\\|\"ZNF365\"\\|\"ZMIZ1\"\\|\"BABAM1\"\\|\"LSP1\"\\|\"ANKLE1\"\\|\"TOPBP1\"\\|\"BCCIP\"\\|\"TP53BP1\"\\|BRCA1\t|BRCA2\t|CHEK2\t|PALB2\t|BRIP1\t|TP53\t|PTEN\t|STK11\t|CDH1\t|ATM\t|BARD1\t|APC\t|MLH1\t|MRE11A\t|MSH2\t|MSH6\t|MUTYH\t|NBN\t|PMS1\t|PMS2\t|RAD50\t|RAD51D\t|RAD51C\t|XRCC2\t|UIMC1\t|FAM175A\t|ERCC4\t|RAD51\t|RAD51B\t|XRCC3\t|FANCA\t|FANCB\t|FANCC\t|FANCD2\t|FANCE\t|FANCF\t|FANCG\t|FANCI\t|FANCL\t|FANCM\t|SLX4\t|CASP8\t|FGFR2\t|TOX3\t|MAP3K1\t|MRPS30\t|SLC4A7\t|NEK10\t|COX11\t|ESR1\t|CDKN2A\t|CDKN2B\t|ANKRD16\t|FBXO18\t|ZNF365\t|ZMIZ1\t|BABAM1\t|LSP1\t|ANKLE1\t|TOPBP1\t|BCCIP\t|TP53BP1  "
  p_filter.c  <- "BRCA1 BRCA2 CHEK2 PALB2 BRIP1 TP53 PTEN STK11 CDH1 ATM BARD1 APC MLH1 MRE11A MSH2 MSH6 MUTYH NBN PMS1 PMS2 RAD50 RAD51D RAD51C XRCC2 UIMC1 FAM175A ERCC4 RAD51 RAD51B XRCC3 FANCA FANCB FANCC FANCD2 FANCE FANCF FANCG FANCI FANCL FANCM SLX4 CASP8 FGFR2 TOX3 MAP3K1 MRPS30 SLC4A7 NEK10 COX11 ESR1 CDKN2A CDKN2B ANKRD16 FBXO18 ZNF365 ZMIZ1 BABAM1 LSP1 ANKLE1 TOPBP1 BCCIP TP53BP1"
  p_tggbf     <- TRUE # TRUE=A bed file will be generated with the intersecting intervals for the target genes. Needed to filter vcf files for target genes before reunning custom snpEff Report 
  p_mail.send <- 0 # 0=FALSE, 1=TRUE ; Indicate whether we want an email sent when the run is finished
  p_only.m.r  <- 0 # Use only Mapped reads to created the corresponding bam files?
  #                  0/n: no, use mapped and unmapped reads; 
  #                  1/y: yes, use only mapped reads; 
  #                 -1/u: the opposite, use only unmapped reads.
  p_bychr     <- 0  # Split the processing by chromosomes (variant calling only, as of April 23, 2013). 
                    # Watch out that some QUAL might be recorded as lower than the value indicated at the bam file
                    # for some misterous reason. See comments inside function var.calling
  p_convert.file.list.pe1        <- TRUE #FALSE #TRUE # Keep as TRUE if you have paired end samples (sampe; p_bwa=2)
  p_convert.file.list.pe2        <- TRUE # This needs to be FALSE for using sam files as input files for the pipeline
  #                                 --and you are re-processing just some steps--, since you will need
  #                                 the input file list 1_sequence.fastq etc converted into the _merged12.sam 
  #                                 IMPORTANT: 
  #                                 Since revision bzr 70'ish (mid Jan'2013), this might need to be enabled as TRUE in both cases 
  #                                 of running the whole pipeline or just some steps, when using short reads paired end (sampe; p_bwa=2)
  
} else if (p_test==2) {
  path_input_absolute <- "0" # Define whether the p_input is absolute or relative
  p_input    <- "test_in2" #"/mnt/magatzem02/tmp/run_sara_293a/dir_in_293a3" #"test_in2" # "../test_in2"  # "test_in"
  p_in.suffix <- "" #"_sequence" # "" # This is the suffix of all input filenames (without extension) used for the pipeline to process
  p_in.ext    <- ".sam" #".fastq" # ".fa" ".sam" ".bam" # This is the .extension of all files used as input for the pipeline to process
  p_output   <- "test_out2" #"/mnt/magatzem02/tmp/run_sara_293a/dir_out_293a3b" #"test_out2" # "../test_out2" # "test_out"
  p_f_my_rs  <- "file_my_rs.txt" # In p_input. Needed by SnpEff to filter for the target genes before the report (well, filter for the potential snp rs codes in those genes)
  p_label    <- "test2_19" #"testrunGATK1" # "testsnpEffCountReads_a" "test-121002" # "test-foo"        # Run Label for output filenames
  p_desc     <- "Testing run
                        test2_19: Testing the count reads: 
                                    New functions to reduce memory consumption (mmap, etc)"
  #                      test2_18: Testing the count reads: 
  #                                  How long does it take with the test set?"
  #                      test2_14c: As in test2_14b, but only with the function to fix cols: 
  #                                  testing the a csv file heavier to test for preformance improvement (right now it was 1 second only)"
  #                      test2_14b: As in test2_14, no changes so far: 
  #                                  testing the new functions to fix INFO fields in splitted columns from csv files from anovar"
  #                      test2_17: Test the splitting of bam in chromosomes prior to variant calling"
  #                      test2_16: Test the param to filter by read base quality (Q) at the variant calling process"
  #                      test2_15: Test the new function to mark duplicates with Picard, prior to removal by samtools rmdup"
  #                      test2_14: Test the new functions to fix INFO fields in splitted columns from csv files from anovar"
  #                      test2_13: Skipped testing the new functions to write the html report (for te time being). 
  #                                Testing samtools remove pcr dups"
  #                      test2_12: Testing the new functions to write the html report. 
  #                                Same params as test2_09. apparently all db are in. And run in parallel again"
  #                      test2_11: Minor typos in var filtering (symlinks) and grep calls post annovar annotations fixed. 
  #                                Same params as test2_09. apparently all db are in. And run in parallel again"
  #                      test2_10: OK. Same params as test2_09, but checking now new annovar version ref files. 
  #                                Some db must be missing"
  #                      test2_09: OK. As test2_08 but with new bed file (tab format) + chr for chromosomes.
  #                                + params$opt$p_se_db_rg instead of hg19 for snpeff. Multithread mode removed to snpEff in order to have stats html file generated again"
  #                      test2_08: As test2_07 but with param a & d from samtools vcftools varFilter as 1 to get max number of candidate variants
  #                                Then, annotate with samtools, and also with SnpSift." 
  #                      test2_07: As test2_06 but with updated code (r92). Playing with Annovar new dbSNP updated files, 132 for the time being  (135, 137)." 
  #                      test2_06: As testset05 but playing with bed files to filter for intersecting intervals from target genes" 
  #                      testset05: As testset4 but going back to test the variant filtering process with samtools, since we get zero results with current default settings
  #                               varFilter -Q10 -d2  -a1 -D10000000 -S1000 ->   72 records in filtered.vcf
  #                               varFilter -Q5  -d1  -a2 -D10000000 -S1000 ->   32 records in filtered.vcf
  #                               varFilter -Q5  -d1  -a1 -D10000000 -S1000 ->  716 records in filtered.vcf
  #                               varFilter -Q5  -d2  -a1 -D10000000 -S1000 ->   74 records in filtered.vcf
  #                               varFilter -Q10 -d2  -a1 -D10000000 -S1000 ->   74 records in filtered.vcf
  #                               varFilter -Q10 -d2  -a2 -D10000000 -S1000 ->   32 records in filtered.vcf
  #                               varFilter -Q1  -d2  -a2 -D10000000 -S1    ->   32 records in filtered.vcf
  #                               varFilter -Q10 -d2  -a2 -D10000000 -S1    ->   32 records in filtered.vcf
  #                               varFilter -Q10 -d10 -a1 -D10000000 -S1000 ->    0 records in filtered.vcf
  #                               varFilter -Q10 -d5  -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
  #                               varFilter -Q1  -d10 -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
  #                               varFilter -Q5  -d10 -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
  #                               varFilter -Q10 -d10 -a2 -D10000000 -S1000 ->    0 records in filtered.vcf
  #                               "
  #                   testset04: As testset3 but playing with bed files to filter for intersecting intervals from target genes" # "Testing the new testset1 from the source merged sam file. See http://ueb.vhir.org/PendentsEVA and http://ueb.vhir.org/sheet3" #"Testing gatk files etc"# Testing The issue in snpEff Count Reads with first set: s_7_m11_149b_merged12_00-01M.sam"
  #                 testset03: As testset2 but variant calling: bcftools without -A, to see if X go away from alternative nucleotides in vcf files" # "Testing the new testset1 from the source merged sam file. See http://ueb.vhir.org/PendentsEVA and http://ueb.vhir.org/sheet3" #"Testing gatk files etc"# Testing The issue in snpEff Count Reads with first set: s_7_m11_149b_merged12_00-01M.sam"
  #    "testset99: As in testset99-1, but with new version of annovar and associated db:
  #             * Annovar from 2013 02 21 (instead of 2012 10).
  #             * 131M Feb 11  2013 hg19_refGeneMrna.fa
  #             *  54M Mar  7  2013 hg19_cosmic63.txt
  #             * 6.2M Mar  7  2013 hg19_cosmic63.txt.idx
  #             *  17K Feb 26  2013 hg19_MT_GRCh37_ensGeneMrna.fa
  #             * 3.2K Feb 26  2013 hg19_MT_GRCh37_ensGene.txt
  #             * 131M Feb 11  2013 hg19_refGeneMrna.fa
  #             *  12M Mar 18  2013 hg19_refGene.txt
  #             *  29M Mar 18  2013 hg19_refLink.txt
  #             * 845M May 25  2012 hg19_snp132NonFlagged.txt.gz
  #             * 846M Oct  1  2011 hg19_snp132.txt.gz
  #             * 1.3G May 25  2012 hg19_snp135NonFlagged.txt.gz
  #             * 1.3G Feb 23  2012 hg19_snp135.txt.gz
  #             * 1.4G Jan  9  2013 hg19_snp137NonFlagged.txt.gz
  #             * 1.4G Jan  7  2013 hg19_snp137.txt.gz
  #             "
  
  #                 testset2: As testset1 but with my_rs.txt added to filter for known snps from target genes" # "Testing the new testset1 from the source merged sam file. See http://ueb.vhir.org/PendentsEVA and http://ueb.vhir.org/sheet3" #"Testing gatk files etc"# Testing The issue in snpEff Count Reads with first set: s_7_m11_149b_merged12_00-01M.sam"
  p_keep     <- TRUE # Enable if run through editor and you want to keep temp files
  p_showc    <- TRUE # Enable if you want to see the commands literally that are run in the command line
  p_st.vf.Q  <-      10 # Samtools Q param & vcftools.pl varFilter Q param: Minimum base quality for a base to be considered: 10 (same for small testing sets). 10 (90% accuracy)|20 (99.9%)|30 (99.99%)|40 (99.999%)|50 (99.9999%). See Wikipedia.".
  p_st.vf.d  <-       1 #(2) Samtools vcftools.pl varFilter d param: Minimum read depth (coverage) to call a SNP: 10-15 (1-2 for small testing sets)".
  p_st.vf.a  <-       1 #(1) Samtools vcftools.pl varFilter a param: Minimum number of alternate bases: 2-3 (1-2 for small testing sets)".
  p_st.vf.D  <-10000000 # Samtools vcftools.pl varFilter D param: Maximum read depth (coverage) to call a SNP: 10000000 (same for small testing sets)".
  p_st.vf.S  <-    1000 # Samtools vcftools.pl varFilter S param: minimum SNP quality: 1000 (same for small testing sets). The smaller, the better (more precise, more quality in the SNP). High values are too permissive."
  #p_filter   <- ""   
  p_filter   <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1:\\|^BRCA1\t\\|^BRCA2\t\\|^CHEK2\t\\|^PALB2\t\\|^BRIP1\t\\|^TP53\t\\|^PTEN\t\\|^STK11\t\\|^CDH1\t\\|^ATM\t\\|^BARD1\t\\|^APC\t\\|^MLH1\t\\|^MRE11A\t\\|^MSH2\t\\|^MSH6\t\\|^MUTYH\t\\|^NBN\t\\|^PMS1\t\\|^PMS2\t\\|^RAD50\t\\|^RAD51D\t\\|^RAD51C\t\\|^XRCC2\t\\|^UIMC1\t\\|^FAM175A\t\\|^ERCC4\t\\|^RAD51\t\\|^RAD51B\t\\|^XRCC3\t\\|^FANCA\t\\|^FANCB\t\\|^FANCC\t\\|^FANCD2\t\\|^FANCE\t\\|^FANCF\t\\|^FANCG\t\\|^FANCI\t\\|^FANCL\t\\|^FANCM\t\\|^SLX4\t\\|^CASP8\t\\|^FGFR2\t\\|^TOX3\t\\|^MAP3K1\t\\|^MRPS30\t\\|^SLC4A7\t\\|^NEK10\t\\|^COX11\t\\|^ESR1\t\\|^CDKN2A\t\\|^CDKN2B\t\\|^ANKRD16\t\\|^FBXO18\t\\|^ZNF365\t\\|^ZMIZ1\t\\|^BABAM1\t\\|^LSP1\t\\|^ANKLE1\t\\|^TOPBP1\t\\|^BCCIP\t\\|^TP53BP1\t\\|\"BRCA1\"\\|\"BRCA2\"\\|\"CHEK2\"\\|\"PALB2\"\\|\"BRIP1\"\\|\"TP53\"\\|\"PTEN\"\\|\"STK11\"\\|\"CDH1\"\\|\"ATM\"\\|\"BARD1\"\\|\"APC\"\\|\"MLH1\"\\|\"MRE11A\"\\|\"MSH2\"\\|\"MSH6\"\\|\"MUTYH\"\\|\"NBN\"\\|\"PMS1\"\\|\"PMS2\"\\|\"RAD50\"\\|\"RAD51D\"\\|\"RAD51C\"\\|\"XRCC2\"\\|\"UIMC1\"\\|\"FAM175A\"\\|\"ERCC4\"\\|\"RAD51\"\\|\"RAD51B\"\\|\"XRCC3\"\\|\"FANCA\"\\|\"FANCB\"\\|\"FANCC\"\\|\"FANCD2\"\\|\"FANCE\"\\|\"FANCF\"\\|\"FANCG\"\\|\"FANCI\"\\|\"FANCL\"\\|\"FANCM\"\\|\"SLX4\"\\|\"CASP8\"\\|\"FGFR2\"\\|\"TOX3\"\\|\"MAP3K1\"\\|\"MRPS30\"\\|\"SLC4A7\"\\|\"NEK10\"\\|\"COX11\"\\|\"ESR1\"\\|\"CDKN2A\"\\|\"CDKN2B\"\\|\"ANKRD16\"\\|\"FBXO18\"\\|\"ZNF365\"\\|\"ZMIZ1\"\\|\"BABAM1\"\\|\"LSP1\"\\|\"ANKLE1\"\\|\"TOPBP1\"\\|\"BCCIP\"\\|\"TP53BP1\"\\|BRCA1\t|BRCA2\t|CHEK2\t|PALB2\t|BRIP1\t|TP53\t|PTEN\t|STK11\t|CDH1\t|ATM\t|BARD1\t|APC\t|MLH1\t|MRE11A\t|MSH2\t|MSH6\t|MUTYH\t|NBN\t|PMS1\t|PMS2\t|RAD50\t|RAD51D\t|RAD51C\t|XRCC2\t|UIMC1\t|FAM175A\t|ERCC4\t|RAD51\t|RAD51B\t|XRCC3\t|FANCA\t|FANCB\t|FANCC\t|FANCD2\t|FANCE\t|FANCF\t|FANCG\t|FANCI\t|FANCL\t|FANCM\t|SLX4\t|CASP8\t|FGFR2\t|TOX3\t|MAP3K1\t|MRPS30\t|SLC4A7\t|NEK10\t|COX11\t|ESR1\t|CDKN2A\t|CDKN2B\t|ANKRD16\t|FBXO18\t|ZNF365\t|ZMIZ1\t|BABAM1\t|LSP1\t|ANKLE1\t|TOPBP1\t|BCCIP\t|TP53BP1  "
  p_filter.c  <- "BRCA1 BRCA2 CHEK2 PALB2 BRIP1 TP53 PTEN STK11 CDH1 ATM BARD1 APC MLH1 MRE11A MSH2 MSH6 MUTYH NBN PMS1 PMS2 RAD50 RAD51D RAD51C XRCC2 UIMC1 FAM175A ERCC4 RAD51 RAD51B XRCC3 FANCA FANCB FANCC FANCD2 FANCE FANCF FANCG FANCI FANCL FANCM SLX4 CASP8 FGFR2 TOX3 MAP3K1 MRPS30 SLC4A7 NEK10 COX11 ESR1 CDKN2A CDKN2B ANKRD16 FBXO18 ZNF365 ZMIZ1 BABAM1 LSP1 ANKLE1 TOPBP1 BCCIP TP53BP1"
  p_tggbf     <- TRUE #TRUE # TRUE=A bed file will be generated with the intersecting intervals for the target genes. Needed to filter vcf files for target genes before reunning custom snpEff Report 
  p_mail.send <- 0 # 0=FALSE, 1=TRUE ; Indicate whether we want an email sent when the run is finished
  p_only.m.r  <- 1 # Use only Mapped reads to created the corresponding bam files?
  #                 0/n: no, use mapped and unmapped reads; 
  #                 1/y: yes, use only mapped reads; 
  #                -1/u: the opposite, use only unmapped reads.
  p_bychr     <- 0  # Split the processing by chromosomes (variant calling only, as of April 23, 2013). 
                    # Watch out that some QUAL might be recorded as lower than the value indicated at the bam file
                    # for some misterous reason. See comments inside function var.calling
  p_convert.file.list.pe1        <- TRUE #FALSE #TRUE # Keep as TRUE if you have paired end samples (sampe; p_bwa=2)
  p_convert.file.list.pe2        <- FALSE # This needs to be FALSE for using sam files as input files for the pipeline
  #                                 --and you are re-processing just some steps--, since you will need
  #                                 the input file list 1_sequence.fastq etc converted into the _merged12.sam 
  #                                 IMPORTANT: 
  #                                 Since revision bzr 70'ish (mid Jan'2013), this might need to be enabled as TRUE in both cases 
  #                                 of running the whole pipeline or just some steps, when using short reads paired end (sampe; p_bwa=2)
  #################################################################
  } else { # p_test     = 0. Run with real data (no testing)
  #################################################################
  path_input_absolute <- "1" # Define whether the p_input is absolute or relative
  
  # PARAMS for dir_out_all from dir_in_test_cr_ind1: Individual 1.
  # -----------------------------------------
  p_input    <- "/mnt/magatzem02/tmp/run_sara_293a/dir_in_test_cr_ind1" # "../dir_in" # "test_in"   # "dir_in"     
  p_output   <- "/mnt/magatzem02/tmp/run_sara_293a/dir_out_all" #../dir_out_293" # "../dir_out_293" # "test_out"   # "dir_out_293"
  p_f_my_rs  <- "file_my_rs.txt" # In p_input. Needed by SnpEff to filter for the target genes before the report (well, filter for the potential snp rs codes in those genes)
  p_label    <-  "sg293_i01_03" # sg293a2b2.snpeff.greped" # "test-121002" # ".sg293_qa_sg3sg4"   # "test-121002" ".sara207_4s4cpu"        # Run Label for output filenames
  p_desc     <- "Individual 1
                      sg293_i01_03: running count reads on ind 1 only, all regions (no bed file in use, param -i disabled). 
                                  Previous run interferred with other processed due (presumably)  to RAM consumption.
                                  Testing count reads splitted in two processes, and now only postprocessing in R run.
                                  package mmap in use at cr.postprocess, and memory.profile functions applied to see memory consumption along the run in a few specific steps"
  #                    sg293_i01_02: running count reads on ind 1 only, all regions (no bed file in use, param -i disabled)
  #                                Process run while sam2bam is being applied to 4 samples, 8Gb RAM consumed by other processes, and 4 cpus already used"
  #                    sg293_i01_01: running count reads on ind 1 only, with only regions from bed file."
  #                    sg293a5_b12: as in sg293a5_b10, but re-run quality control, as it was missing for some reason for individual 6"
  
# # PARAMS for dir_out_293a5 Individuals 5 & 6.
# # -----------------------------------------
#   p_input    <- "/mnt/magatzem02/tmp/run_sara_293a/dir_in_293a5" # "../dir_in" # "test_in"   # "dir_in"     
#   p_output   <- "/mnt/magatzem02/tmp/run_sara_293a/dir_out_293a5" #../dir_out_293" # "../dir_out_293" # "test_out"	 # "dir_out_293"
#   p_f_my_rs  <- "file_my_rs.txt" # In p_input. Needed by SnpEff to filter for the target genes before the report (well, filter for the potential snp rs codes in those genes)
#   p_label    <-  "sg293a5_b13" # sg293a2b2.snpeff.greped" # "test-121002" # ".sg293_qa_sg3sg4"   # "test-121002" ".sara207_4s4cpu"        # Run Label for output filenames
#   p_desc     <- "Individuals 5 & 6.
#                       sg293a5_b13: as in previous, but re-running ssfii (filter intersecting intervals) to generate the my_genes snpEff reports, etc"
  #                    sg293a5_b12: as in sg293a5_b10, but re-run quality control, as it was missing for some reason for individual 6"
  #                    sg293a5_b11: as in sg293a5_b10, As in sg293a4b_06 but testing functions in snpEff and snpSift related with dbSnp"
  #                    sg293a5_b10: as in sg293a5_b09, but running only grep post snpEff report (with r101)"
  #                    sg293a5_b09: as in sg293a5_b08, but after hardlink to f.vcf is created, adn thus, var filtering not called again. Only Annotation with annovar"
  #                    sg293a5_b08: as in sg293a5_b07, but focused on annotations with annovar. Using bazaar revision100."
  #                    sg293a5_b07: as in sg293a5_b6. SnpEff reports (for all and for target-genes only). Using bazaar revision98."
  #                    sg293a5_b6: as in sg293a5_b5. After count reads done. Following with Variant calling and filtering"
  #                    sg293a5_b5: as in sg293a5_b4 but count reads failed due to lack of enough RAM (other process aside of this one eat it all)"

# PARAMS for dir_out_293a4b. Individuals 3 & 4.
# -----------------------------------------
#   p_input    <- "/mnt/magatzem02/tmp/run_sara_293a/dir_in_293a4" # "../dir_in" # "test_in"   # "dir_in"     
#      p_output   <- "/mnt/magatzem02/tmp/run_sara_293a/dir_out_293a4b" #../dir_out_293" # "../dir_out_293" # "test_out"   # "dir_out_293"
#      p_f_my_rs  <- "file_my_rs.txt" # In p_input. Needed by SnpEff to filter for the target genes before the report (well, filter for the potential snp rs codes in those genes)
#      p_label    <-  "sg293a4b_02" # sg293a2b2.snpeff.greped" # "test-121002" # ".sg293_qa_sg3sg4"   # "test-121002" ".sara207_4s4cpu"        # Run Label for output filenames
#      p_desc     <- "Individuals 3 & 4.
#                          sg293a4b_02: bed files ok, redoing var calling, var filtering, SnpEff reports (for all and for target-genes only). Using bazaar revision98."
#   
#   #  p_input     <- "/backups_disk_03/tmp/bam_mv311" # "../dir_in" # "test_in"   # "dir_in"     
  p_in.suffix <- "_sequence" #  "_sequence" # "" # This is the suffix of all input filenames (without extension) used for the pipeline to process
  p_in.ext    <-  ".fastq" #".fastq" # ".fa" ".sam" ".bam" # This is the .extension of all files used as input for the pipeline to process
#  p_output   <- "/backups_disk_03/tmp/bam_mv311" #"/home/xavi/Estudis/eva_bowtie/dir_out" #../dir_out_293" # "../dir_out_293" # "test_out"   # "dir_out_293"
#  p_label    <-   "mv311c3" #sg293a3_test_count_reads" # ".sg293a3_ind7_mc15g3seq" # mc=minimum cover; g3: 3rd filter version for the grep; "test-121002" # ".sg293_qa_sg3sg4"   # "test-121002" ".sara207_4s4cpu"        # Run Label for output filenames
                  # p_desc = Description in longer format of the run. DEscribe that for you to understand in the future what were the conditions and params of this run. 
#  p_desc     <-   "c3: processat d'un fastq de prova fins a sam i bam amb el pipeline eva per a ratoli rn4."
  p_keep     <- FALSE # Enable if run through editor and you want to keep temp files
  p_showc    <- TRUE #TRUE # Enable if you want to see the commands literally that are run in the command line
  p_st.vf.Q  <-      10 # Samtools vcftools.pl varFilter Q param: Minimum base quality for a base to be considered: 10 (same for small testing sets). 10 (90% accuracy)|20 (99.9%)|30 (99.99%)|40 (99.999%)|50 (99.9999%). See Wikipedia.".
  p_st.vf.d  <-      10 # Samtools vcftools.pl varFilter d param: Minimum read depth (coverage) to call a SNP: 10-15 (1-2 for small testing sets)".
  p_st.vf.a  <-       2 # Samtools vcftools.pl varFilter a param: Minimum number of alternate bases: 2-3 (1-2 for small testing sets)".
  p_st.vf.D  <-10000000 # Samtools vcftools.pl varFilter D param: Maximum read depth (coverage) to call a SNP: 10000000 (same for small testing sets)".
  p_st.vf.S  <-    1000 # Samtools vcftools.pl varFilter S param: minimum SNP quality: 1000 (same for small testing sets). The smaller, the better (more precise, more quality in the SNP). High values are too permissive."
  #  p_filter   <- "BRCA"
#  p_filter   <- "BRCA1\\|BRCA2\\|CHEK2\\|PALB2\\|BRIP1\\|TP53\\|PTEN\\|STK11\\|CDH1\\|ATM\\|BARD1\\|APC\\|MLH1\\|MRE11A\\|MSH2\\|MSH6\\|MUTYH\\|NBN\\|PMS1\\|PMS2\\|RAD50\\|RAD51D\\|RAD51C\\|XRCC2\\|UIMC1\\|FAM175A\\|ERCC4\\|RAD51\\|RAD51B\\|XRCC3\\|FANCA\\|FANCB\\|FANCC\\|FANCD2\\|FANCE\\|FANCF\\|FANCG\\|FANCI\\|FANCL\\|FANCM\\|SLX4\\|CASP8\\|FGFR2\\|TOX3\\|MAP3K1\\|MRPS30\\|SLC4A7\\|NEK10\\|COX11\\|ESR1\\|CDKN2A\\|CDKN2B\\|ANKRD16\\|FBXO18\\|ZNF365\\|ZMIZ1\\|BABAM1\\|LSP1\\|ANKLE1\\|TOPBP1\\|BCCIP\\|TP53BP1"            
  # for p_filter, keep in mind that we want to find 'foo1', and there is one 'foo10', we need to indicate somehow to grep for the whole word
  # and that's why we grep for "foo1|", "foo1:", "\"foo\"" anywhere, or "foo\t" at the begining of the line, since the "|" is the character splitter in snpEff results, 
  # ":" in annovar results, and at the begining of the string before a tab (\t) in *snpEff_genes.txt  results file.
  # We are also doing a system (gnu) grep with a -P flag (to indicate to read PERL-like syntax) for "foo\t" at teh snpEff.txt output files wich are tab deliimted.
#  p_filter   <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1"  
#  p_filter   <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1:\\|^BRCA1\t\\|^BRCA2\t\\|^CHEK2\t\\|^PALB2\t\\|^BRIP1\t\\|^TP53\t\\|^PTEN\t\\|^STK11\t\\|^CDH1\t\\|^ATM\t\\|^BARD1\t\\|^APC\t\\|^MLH1\t\\|^MRE11A\t\\|^MSH2\t\\|^MSH6\t\\|^MUTYH\t\\|^NBN\t\\|^PMS1\t\\|^PMS2\t\\|^RAD50\t\\|^RAD51D\t\\|^RAD51C\t\\|^XRCC2\t\\|^UIMC1\t\\|^FAM175A\t\\|^ERCC4\t\\|^RAD51\t\\|^RAD51B\t\\|^XRCC3\t\\|^FANCA\t\\|^FANCB\t\\|^FANCC\t\\|^FANCD2\t\\|^FANCE\t\\|^FANCF\t\\|^FANCG\t\\|^FANCI\t\\|^FANCL\t\\|^FANCM\t\\|^SLX4\t\\|^CASP8\t\\|^FGFR2\t\\|^TOX3\t\\|^MAP3K1\t\\|^MRPS30\t\\|^SLC4A7\t\\|^NEK10\t\\|^COX11\t\\|^ESR1\t\\|^CDKN2A\t\\|^CDKN2B\t\\|^ANKRD16\t\\|^FBXO18\t\\|^ZNF365\t\\|^ZMIZ1\t\\|^BABAM1\t\\|^LSP1\t\\|^ANKLE1\t\\|^TOPBP1\t\\|^BCCIP\t\\|^TP53BP1\t"  
#wrong#  p_filter   <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1:\\|^BRCA1\t\\|^BRCA2\t\\|^CHEK2\t\\|^PALB2\t\\|^BRIP1\t\\|^TP53\t\\|^PTEN\t\\|^STK11\t\\|^CDH1\t\\|^ATM\t\\|^BARD1\t\\|^APC\t\\|^MLH1\t\\|^MRE11A\t\\|^MSH2\t\\|^MSH6\t\\|^MUTYH\t\\|^NBN\t\\|^PMS1\t\\|^PMS2\t\\|^RAD50\t\\|^RAD51D\t\\|^RAD51C\t\\|^XRCC2\t\\|^UIMC1\t\\|^FAM175A\t\\|^ERCC4\t\\|^RAD51\t\\|^RAD51B\t\\|^XRCC3\t\\|^FANCA\t\\|^FANCB\t\\|^FANCC\t\\|^FANCD2\t\\|^FANCE\t\\|^FANCF\t\\|^FANCG\t\\|^FANCI\t\\|^FANCL\t\\|^FANCM\t\\|^SLX4\t\\|^CASP8\t\\|^FGFR2\t\\|^TOX3\t\\|^MAP3K1\t\\|^MRPS30\t\\|^SLC4A7\t\\|^NEK10\t\\|^COX11\t\\|^ESR1\t\\|^CDKN2A\t\\|^CDKN2B\t\\|^ANKRD16\t\\|^FBXO18\t\\|^ZNF365\t\\|^ZMIZ1\t\\|^BABAM1\t\\|^LSP1\t\\|^ANKLE1\t\\|^TOPBP1\t\\|^BCCIP\t\\|^TP53BP1\t\\|\\"BRCA1\\"\\|\\"BRCA2\\"\\|\\"CHEK2\\"\\|\\"PALB2\\"\\|\\"BRIP1\\"\\|\\"TP53\\"\\|\\"PTEN\\"\\|\\"STK11\\"\\|\\"CDH1\\"\\|\\"ATM\\"\\|\\"BARD1\\"\\|\\"APC\\"\\|\\"MLH1\\"\\|\\"MRE11A\\"\\|\\"MSH2\\"\\|\\"MSH6\\"\\|\\"MUTYH\\"\\|\\"NBN\\"\\|\\"PMS1\\"\\|\\"PMS2\\"\\|\\"RAD50\\"\\|\\"RAD51D\\"\\|\\"RAD51C\\"\\|\\"XRCC2\\"\\|\\"UIMC1\\"\\|\\"FAM175A\\"\\|\\"ERCC4\\"\\|\\"RAD51\\"\\|\\"RAD51B\\"\\|\\"XRCC3\\"\\|\\"FANCA\\"\\|\\"FANCB\\"\\|\\"FANCC\\"\\|\\"FANCD2\\"\\|\\"FANCE\\"\\|\\"FANCF\\"\\|\\"FANCG\\"\\|\\"FANCI\\"\\|\\"FANCL\\"\\|\\"FANCM\\"\\|\\"SLX4\\"\\|\\"CASP8\\"\\|\\"FGFR2\\"\\|\\"TOX3\\"\\|\\"MAP3K1\\"\\|\\"MRPS30\\"\\|\\"SLC4A7\\"\\|\\"NEK10\\"\\|\\"COX11\\"\\|\\"ESR1\\"\\|\\"CDKN2A\\"\\|\\"CDKN2B\\"\\|\\"ANKRD16\\"\\|\\"FBXO18\\"\\|\\"ZNF365\\"\\|\\"ZMIZ1\\"\\|\\"BABAM1\\"\\|\\"LSP1\\"\\|\\"ANKLE1\\"\\|\\"TOPBP1\\"\\|\\"BCCIP\\"\\|\\"TP53BP1\\""
#  p_filter   <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1:\\|^BRCA1\t\\|^BRCA2\t\\|^CHEK2\t\\|^PALB2\t\\|^BRIP1\t\\|^TP53\t\\|^PTEN\t\\|^STK11\t\\|^CDH1\t\\|^ATM\t\\|^BARD1\t\\|^APC\t\\|^MLH1\t\\|^MRE11A\t\\|^MSH2\t\\|^MSH6\t\\|^MUTYH\t\\|^NBN\t\\|^PMS1\t\\|^PMS2\t\\|^RAD50\t\\|^RAD51D\t\\|^RAD51C\t\\|^XRCC2\t\\|^UIMC1\t\\|^FAM175A\t\\|^ERCC4\t\\|^RAD51\t\\|^RAD51B\t\\|^XRCC3\t\\|^FANCA\t\\|^FANCB\t\\|^FANCC\t\\|^FANCD2\t\\|^FANCE\t\\|^FANCF\t\\|^FANCG\t\\|^FANCI\t\\|^FANCL\t\\|^FANCM\t\\|^SLX4\t\\|^CASP8\t\\|^FGFR2\t\\|^TOX3\t\\|^MAP3K1\t\\|^MRPS30\t\\|^SLC4A7\t\\|^NEK10\t\\|^COX11\t\\|^ESR1\t\\|^CDKN2A\t\\|^CDKN2B\t\\|^ANKRD16\t\\|^FBXO18\t\\|^ZNF365\t\\|^ZMIZ1\t\\|^BABAM1\t\\|^LSP1\t\\|^ANKLE1\t\\|^TOPBP1\t\\|^BCCIP\t\\|^TP53BP1\t\\|\"BRCA1\"\\|\"BRCA2\"\\|\"CHEK2\"\\|\"PALB2\"\\|\"BRIP1\"\\|\"TP53\"\\|\"PTEN\"\\|\"STK11\"\\|\"CDH1\"\\|\"ATM\"\\|\"BARD1\"\\|\"APC\"\\|\"MLH1\"\\|\"MRE11A\"\\|\"MSH2\"\\|\"MSH6\"\\|\"MUTYH\"\\|\"NBN\"\\|\"PMS1\"\\|\"PMS2\"\\|\"RAD50\"\\|\"RAD51D\"\\|\"RAD51C\"\\|\"XRCC2\"\\|\"UIMC1\"\\|\"FAM175A\"\\|\"ERCC4\"\\|\"RAD51\"\\|\"RAD51B\"\\|\"XRCC3\"\\|\"FANCA\"\\|\"FANCB\"\\|\"FANCC\"\\|\"FANCD2\"\\|\"FANCE\"\\|\"FANCF\"\\|\"FANCG\"\\|\"FANCI\"\\|\"FANCL\"\\|\"FANCM\"\\|\"SLX4\"\\|\"CASP8\"\\|\"FGFR2\"\\|\"TOX3\"\\|\"MAP3K1\"\\|\"MRPS30\"\\|\"SLC4A7\"\\|\"NEK10\"\\|\"COX11\"\\|\"ESR1\"\\|\"CDKN2A\"\\|\"CDKN2B\"\\|\"ANKRD16\"\\|\"FBXO18\"\\|\"ZNF365\"\\|\"ZMIZ1\"\\|\"BABAM1\"\\|\"LSP1\"\\|\"ANKLE1\"\\|\"TOPBP1\"\\|\"BCCIP\"\\|\"TP53BP1\""
  p_filter    <- "BRCA1|\\|BRCA2|\\|CHEK2|\\|PALB2|\\|BRIP1|\\|TP53|\\|PTEN|\\|STK11|\\|CDH1|\\|ATM|\\|BARD1|\\|APC|\\|MLH1|\\|MRE11A|\\|MSH2|\\|MSH6|\\|MUTYH|\\|NBN|\\|PMS1|\\|PMS2|\\|RAD50|\\|RAD51D|\\|RAD51C|\\|XRCC2|\\|UIMC1|\\|FAM175A|\\|ERCC4|\\|RAD51|\\|RAD51B|\\|XRCC3|\\|FANCA|\\|FANCB|\\|FANCC|\\|FANCD2|\\|FANCE|\\|FANCF|\\|FANCG|\\|FANCI|\\|FANCL|\\|FANCM|\\|SLX4|\\|CASP8|\\|FGFR2|\\|TOX3|\\|MAP3K1|\\|MRPS30|\\|SLC4A7|\\|NEK10|\\|COX11|\\|ESR1|\\|CDKN2A|\\|CDKN2B|\\|ANKRD16|\\|FBXO18|\\|ZNF365|\\|ZMIZ1|\\|BABAM1|\\|LSP1|\\|ANKLE1|\\|TOPBP1|\\|BCCIP|\\|TP53BP1|\\|BRCA1:\\|BRCA2:\\|CHEK2:\\|PALB2:\\|BRIP1:\\|TP53:\\|PTEN:\\|STK11:\\|CDH1:\\|ATM:\\|BARD1:\\|APC:\\|MLH1:\\|MRE11A:\\|MSH2:\\|MSH6:\\|MUTYH:\\|NBN:\\|PMS1:\\|PMS2:\\|RAD50:\\|RAD51D:\\|RAD51C:\\|XRCC2:\\|UIMC1:\\|FAM175A:\\|ERCC4:\\|RAD51:\\|RAD51B:\\|XRCC3:\\|FANCA:\\|FANCB:\\|FANCC:\\|FANCD2:\\|FANCE:\\|FANCF:\\|FANCG:\\|FANCI:\\|FANCL:\\|FANCM:\\|SLX4:\\|CASP8:\\|FGFR2:\\|TOX3:\\|MAP3K1:\\|MRPS30:\\|SLC4A7:\\|NEK10:\\|COX11:\\|ESR1:\\|CDKN2A:\\|CDKN2B:\\|ANKRD16:\\|FBXO18:\\|ZNF365:\\|ZMIZ1:\\|BABAM1:\\|LSP1:\\|ANKLE1:\\|TOPBP1:\\|BCCIP:\\|TP53BP1:\\|^BRCA1\t\\|^BRCA2\t\\|^CHEK2\t\\|^PALB2\t\\|^BRIP1\t\\|^TP53\t\\|^PTEN\t\\|^STK11\t\\|^CDH1\t\\|^ATM\t\\|^BARD1\t\\|^APC\t\\|^MLH1\t\\|^MRE11A\t\\|^MSH2\t\\|^MSH6\t\\|^MUTYH\t\\|^NBN\t\\|^PMS1\t\\|^PMS2\t\\|^RAD50\t\\|^RAD51D\t\\|^RAD51C\t\\|^XRCC2\t\\|^UIMC1\t\\|^FAM175A\t\\|^ERCC4\t\\|^RAD51\t\\|^RAD51B\t\\|^XRCC3\t\\|^FANCA\t\\|^FANCB\t\\|^FANCC\t\\|^FANCD2\t\\|^FANCE\t\\|^FANCF\t\\|^FANCG\t\\|^FANCI\t\\|^FANCL\t\\|^FANCM\t\\|^SLX4\t\\|^CASP8\t\\|^FGFR2\t\\|^TOX3\t\\|^MAP3K1\t\\|^MRPS30\t\\|^SLC4A7\t\\|^NEK10\t\\|^COX11\t\\|^ESR1\t\\|^CDKN2A\t\\|^CDKN2B\t\\|^ANKRD16\t\\|^FBXO18\t\\|^ZNF365\t\\|^ZMIZ1\t\\|^BABAM1\t\\|^LSP1\t\\|^ANKLE1\t\\|^TOPBP1\t\\|^BCCIP\t\\|^TP53BP1\t\\|\"BRCA1\"\\|\"BRCA2\"\\|\"CHEK2\"\\|\"PALB2\"\\|\"BRIP1\"\\|\"TP53\"\\|\"PTEN\"\\|\"STK11\"\\|\"CDH1\"\\|\"ATM\"\\|\"BARD1\"\\|\"APC\"\\|\"MLH1\"\\|\"MRE11A\"\\|\"MSH2\"\\|\"MSH6\"\\|\"MUTYH\"\\|\"NBN\"\\|\"PMS1\"\\|\"PMS2\"\\|\"RAD50\"\\|\"RAD51D\"\\|\"RAD51C\"\\|\"XRCC2\"\\|\"UIMC1\"\\|\"FAM175A\"\\|\"ERCC4\"\\|\"RAD51\"\\|\"RAD51B\"\\|\"XRCC3\"\\|\"FANCA\"\\|\"FANCB\"\\|\"FANCC\"\\|\"FANCD2\"\\|\"FANCE\"\\|\"FANCF\"\\|\"FANCG\"\\|\"FANCI\"\\|\"FANCL\"\\|\"FANCM\"\\|\"SLX4\"\\|\"CASP8\"\\|\"FGFR2\"\\|\"TOX3\"\\|\"MAP3K1\"\\|\"MRPS30\"\\|\"SLC4A7\"\\|\"NEK10\"\\|\"COX11\"\\|\"ESR1\"\\|\"CDKN2A\"\\|\"CDKN2B\"\\|\"ANKRD16\"\\|\"FBXO18\"\\|\"ZNF365\"\\|\"ZMIZ1\"\\|\"BABAM1\"\\|\"LSP1\"\\|\"ANKLE1\"\\|\"TOPBP1\"\\|\"BCCIP\"\\|\"TP53BP1\"\\|BRCA1\t|BRCA2\t|CHEK2\t|PALB2\t|BRIP1\t|TP53\t|PTEN\t|STK11\t|CDH1\t|ATM\t|BARD1\t|APC\t|MLH1\t|MRE11A\t|MSH2\t|MSH6\t|MUTYH\t|NBN\t|PMS1\t|PMS2\t|RAD50\t|RAD51D\t|RAD51C\t|XRCC2\t|UIMC1\t|FAM175A\t|ERCC4\t|RAD51\t|RAD51B\t|XRCC3\t|FANCA\t|FANCB\t|FANCC\t|FANCD2\t|FANCE\t|FANCF\t|FANCG\t|FANCI\t|FANCL\t|FANCM\t|SLX4\t|CASP8\t|FGFR2\t|TOX3\t|MAP3K1\t|MRPS30\t|SLC4A7\t|NEK10\t|COX11\t|ESR1\t|CDKN2A\t|CDKN2B\t|ANKRD16\t|FBXO18\t|ZNF365\t|ZMIZ1\t|BABAM1\t|LSP1\t|ANKLE1\t|TOPBP1\t|BCCIP\t|TP53BP1  "
  p_filter.c  <- "BRCA1 BRCA2 CHEK2 PALB2 BRIP1 TP53 PTEN STK11 CDH1 ATM BARD1 APC MLH1 MRE11A MSH2 MSH6 MUTYH NBN PMS1 PMS2 RAD50 RAD51D RAD51C XRCC2 UIMC1 FAM175A ERCC4 RAD51 RAD51B XRCC3 FANCA FANCB FANCC FANCD2 FANCE FANCF FANCG FANCI FANCL FANCM SLX4 CASP8 FGFR2 TOX3 MAP3K1 MRPS30 SLC4A7 NEK10 COX11 ESR1 CDKN2A CDKN2B ANKRD16 FBXO18 ZNF365 ZMIZ1 BABAM1 LSP1 ANKLE1 TOPBP1 BCCIP TP53BP1"
  p_tggbf     <- TRUE # TRUE # TRUE=A bed file will be generated with the intersecting intervals for the target genes (.my_genes.bed). Needed to filter vcf files for target genes before reunning custom snpEff Report 
  p_mail.send <- 1 # 0=FALSE, 1=TRUE ; Indicate whether we want an email sent when the run is finished
  p_only.m.r  <- 1 # Use only Mapped reads to created the corresponding bam files?
                      # 0/n: no, use mapped and unmapped reads; 
                      # 1/y: yes, use only mapped reads; 
                      # -1/u: the opposite, use only unmapped reads.
  p_bychr     <- 0  # Split the processing by chromosomes (variant calling only, as of April 23, 2013). 
                    # Watch out that some QUAL might be recorded as lower than the value indicated at the bam file
                    # for some misterous reason. See comments inside function var.calling
  p_convert.file.list.pe1        <- TRUE #FALSE #TRUE # Keep as TRUE if you have paired end samples (sampe; p_bwa=2)
  p_convert.file.list.pe2        <- TRUE # This needs to be FALSE for using sam files as input files for the pipeline
  #                                 --and you are re-processing just some steps--, since you will need
  #                                 the input file list 1_sequence.fastq etc converted into the _merged12.sam 
  #                                 IMPORTANT: 
  #                                 Since revision bzr 70'ish (mid Jan'2013), this might need to be enabled as TRUE in both cases 
  #                                 of running the whole pipeline or just some steps, when using short reads paired end (sampe; p_bwa=2)
}
###########################
# Common params for test runs and non-test runs 
p_genedata  <- TRUE # TRUE # TRUE=A txt file will be generated with the data related to the target genes
p_genver    <- "hg19" # hg19 is the only one supported throughout the whole pipeline still, as of January 2013.
p_se_db_rg  <- "GRCh37.66" # GRCh37.66 is the only one supported for SnpEff v3.1 throughout the whole pipeline still, as of March 2013.
p_index     <- FALSE #FALSE # TRUE         
p_log       <- TRUE        
p_dbsnp     <- "137" # e.g. 132 for dbsnp132. You need to download first the corresponding dbsnp files for annovar.
p_summarize <- TRUE  
p_snpeff.of <- "vcf" # Output format for snpEff. Possible values: txt, vcf, gatk, bed, bedAnn (txt will be deprecated, but it can be ocasionally useful still in the meantime)
p_cpus      <- 4             
p_tmp_dir   <- "/mnt/magatzem03/tmp/tmp_dir/"
p_parallel  <- TRUE #FALSE #FALSE #TRUE # Do you want to allow running some parallelized processes at all? (which ones will be specified elsewhere in the code)
p_bwa       <- 2          # Algorythm for mapping with bwa - http://bio-bwa.sourceforge.net/bwa.shtml
                        # 1: bwa aln      + samse  (short reads, single ends, low errors);
                        # 2: bwa aln (x2) + sampe  (short reads, paired ends, low errors);
                        # 3: bwa bwasw             (longer reads, single end only) 

# Reporting by email at the end of the run
p_from <- "xavier.depdedro@vhir.org"
p_to <- "xdpedro@ir.vhebron.net"
p_subject <- paste("EVA run at MAINHEAD finished - ", p_label, sep="")
p_body <- paste(p_subject, " - from ", program_ueb," - See some log information attached", sep="")                   
p_smtp="smtp.ir.vhebron.net"



# Set flags as ON (TRUE) or OFF (FALSE) for all processes from function:
#  wrapper.sequential (wseq)
#----------------------------------
#####
runParam <- FALSE #######################
p_map.on.reference.genome.sequential.mt     <- runParam # In case we run the mapping sequentially, in multi-thread mode for each sample, with as many threads as cpus indicated in the input params (opt$cpus) 

    # ---------------------------------
    # In general, we don't want to run the mapping parallelizing by samples (many samples at once)
    # since they might be some overlapping at the process of creating some intermediary files in the folder
    # where the reference genome is.
    p_map.on.reference.genome.parallel       <- FALSE #runParam #  The opposite to map in sequential mode (when we need to map)
    # ---------------------------------

# Set all params inside a list, so that it's easier to send from main to functions
# wseq : for function wrapper.sequential
params_wseq <- list()
params_wseq <- list(
  p_map.on.reference.genome.sequential.mt  = p_map.on.reference.genome.sequential.mt, 
  p_map.on.reference.genome.parallel    = p_map.on.reference.genome.parallel
)


# 8b. Set flags as ON (TRUE) or OFF (FALSE) for all processes from function:
#  wrapper2.parallelizable.per.sample (w2pps)
#----------------------------------
# p_map.on.reference.genome.parallel  is not defined here but in the previous chunk
#####
runParam <- FALSE #######################
####
p_quality.control             <- runParam
p_bowtie2sam                  <- runParam
#####
runParam <- FALSE #######################
####
p_sam2bam.and.sort		        <- runParam
p_index.bam.file              <- runParam
                  # This step might not be needed, and in addition, it breaks variant calling!
                  p_samtools.fixmate            <- FALSE
                  # Keep p_samtools.fixmate as FALSE until further notice 
p_picard.mark.dup             <- runParam # Testing the function first
p_remove.pcr.dup  	          <- runParam # Not using the mark duplicates output yet (Testing phase still)
p_stats			                  <- runParam
#####
runParam <- FALSE #######################
####
p_gatk.sortbyref              <- FALSE # runParam # Not working properly yet. Error message: Can not open temporary file 1080: Too many open files at /home/ueb/Data/gatk-data-2.3/SortByRef.pl line 95, <$INPUT> line 50737751.
p_gatk.local.realign.step1    <- FALSE # TRUE # FALSE # runParam
p_gatk.local.realign.step2    <- FALSE # runParam # Unfinished work
p_gatk.local.realign.step3    <- FALSE # runParam # Unfinished work
#####
runParam <- FALSE #######################
####
p_variant.calling		          <- runParam
p_variant.filtering		        <- runParam
p_gatk.combine.vcfs           <- FALSE # runParam # Non-started work (place holder only)
p_convert2vcf4		            <- runParam
p_variant.annotation.geneb	  <- runParam
p_variant.annotation.regionb	<- FALSE # runParam # Non-started work (place holder only)
p_variant.annotation.filterb	<- runParam
p_variant.annotation.summarize<- runParam
#####
runParam <- FALSE #######################
####
p_grep.variants  	            <- runParam
p_variant.annotation.s.fixcols<- runParam
p_visualize.variants		      <- FALSE # runParam # Non-started work (place holder only)
#####
runParam <- FALSE #######################
####
p_variant.fii.pre.snpeff      <- runParam
p_variant.filter.pre.snpeff   <- runParam 
p_variant.dbsnp.pre.snpeff    <- FALSE # runParam # Non-started work (place holder only)
p_grep.pre.snpeff.report      <- FALSE # runParam # Not running properly yet
p_variant.eff.report          <- runParam
p_grep.post.snpeff.report     <- runParam
p_snpeff.count.reads          <- runParam 
#####
runParam <- TRUE #######################
####
p_snpeff.cr.postprocess       <- runParam
#####
runParam <- FALSE #######################
####



# Set all params inside a list, so that it's easier to send from main to functions
# w2pps : for function wrapper2.parallelizable.per.sample
#params_w2pps <- list()
params_w2pps <- list(
  p_map.on.reference.genome.parallel  = p_map.on.reference.genome.parallel,
  p_quality.control                   = p_quality.control,
  p_convert.file.list.pe1             = p_convert.file.list.pe1,
  p_convert.file.list.pe2             = p_convert.file.list.pe2,
  p_bowtie2sam                        = p_bowtie2sam,
  p_sam2bam.and.sort                  = p_sam2bam.and.sort,
  p_index.bam.file                    = p_index.bam.file,
  p_samtools.fixmate                  = p_samtools.fixmate,
  p_picard.mark.dup                   = p_picard.mark.dup,
  p_remove.pcr.dup                    = p_remove.pcr.dup,
  p_gatk.sortbyref                    = p_gatk.sortbyref,
  p_gatk.local.realign.step1          = p_gatk.local.realign.step1,
  p_gatk.local.realign.step2          = p_gatk.local.realign.step2,
  p_gatk.local.realign.step3          = p_gatk.local.realign.step3,
  p_stats                             = p_stats,
  p_variant.calling                   = p_variant.calling,
  p_variant.filtering                 = p_variant.filtering,
  p_gatk.combine.vcfs                 = p_gatk.combine.vcfs,
  p_convert2vcf4                      = p_convert2vcf4,
  p_variant.annotation.geneb          = p_variant.annotation.geneb,
  p_variant.annotation.regionb        = p_variant.annotation.regionb,
  p_variant.annotation.filterb        = p_variant.annotation.filterb,
  p_variant.annotation.summarize      = p_variant.annotation.summarize,
  p_grep.variants                     = p_grep.variants,
  p_variant.annotation.s.fixcols      = p_variant.annotation.s.fixcols,
  p_visualize.variants                = p_visualize.variants,
  p_variant.fii.pre.snpeff            = p_variant.fii.pre.snpeff,
  p_variant.filter.pre.snpeff         = p_variant.filter.pre.snpeff,
  p_variant.dbsnp.pre.snpeff          = p_variant.dbsnp.pre.snpeff,
  p_grep.pre.snpeff.report            = p_grep.pre.snpeff.report,
  p_variant.eff.report                = p_variant.eff.report,
  p_grep.post.snpeff.report           = p_grep.post.snpeff.report,
  p_snpeff.count.reads                = p_snpeff.count.reads,
  p_snpeff.cr.postprocess             = p_snpeff.cr.postprocess
)

# 9b. Set flags as ON (TRUE) or OFF (FALSE) for all processes from function:
#  wrapper2.parallelizable.final (w2pf)
#----------------------------------
#####
runParam <- FALSE #######################
####
p_build.html.report  <- FALSE # runParam # Non-started work (place holder only)


# 9b. Distribute final Parallelizable calculation 
# w2pf : for function wrapper2.parallelizable.final
params_w2pf <- list()
params_w2pf <- list(
  # p_foo_w2pf = p_foo_w2pf, # dummy param to attempt to prevent error message "Error in cut.default(i, breaks) : 'breaks' are not unique" https://bugs.r-project.org/bugzilla3/show_bug.cgi?id=14898
  p_build.html.report             = p_build.html.report
)


# Define path params for all runs
#----------------------------------
# p_server = Choose machine where to get the paths for
# 1 for MainHead,
# 2 for B52,
p_server <- 1 # Set the server number (see codes above)

if (p_server==1) { # MainHead server
  # A. Reference Data file paths
    # A.i) Human Genome
      path_genome1 = "/home/ueb/Data/Data_Genomes/hg19/hg19.fa"  # chromosome names with prefix "chr"
      path_genome2 = "/home/ueb/Data/Data_Genomes/hg19_Broad_Reference_Genome/Homo_sapiens_assembly19.fasta" # chromosome names without prefix "chr"
      path_genome3 = "/home/ueb/Data/gatk-data-2.3/hg19/ucsc.hg19.fasta"
      path_genome4 = "/home/ueb/Data/Data_Genomes/forGATK/hg19_rCRSchrm.fa" # from http://epigenome.usc.edu/publicationdata/bissnp2011/utilies.html
      path_genome5 = "/home/ueb/Data/Data_Genomes/hg19/chr/hg19gatk.fa"  # chromosome names with prefix "chr" and in the order for gatk indicated in http://seqanswers.com/wiki/How-to/exome_analysis
    path_genome = path_genome1
    # A.ii) dbSNP
#      path_dbSNP1 = "/home/ueb/Data/dbSNP/dbsnp132_20101103.vcf" # chromosome names without prefix "chr"
      path_dbSNP1 = "/home/ueb/Data/dbSNP/dbsnp132_20101103_gatk.vcf" # chromosome names with the prefix "chr" being added by Xavi (Jan 23, 2013)
      path_dbSNP2 = "/home/ueb/Data/Data_Genomes/hg19_Broad_Reference_Genome/dbsnp_135_human_9606_v4.0_00-All.vcf" # chromosome names without prefix "chr"
      path_dbSNP3 = "/home/ueb/Data/gatk-data-2.3/hg19/dbsnp_137.hg19.vcf"
      path_dbSNP4 = "/home/ueb/Data/Data_Genomes/forGATK/dbsnp_135.hg19.sort.vcf"
      path_dbSNP5 = "/home/ueb/Data/dbSNP/dbSnp_137.vcf" # (v 18-JUN-2012, dbSnp137) Taken and renamed from ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/00-All.vcf.gz
                      # $ wget -O - ftp://ftp.ncbi.nih.gov/snp/organisms/human_9606/VCF/00-All.vcf.gz | zcat | head
                      ##fileDate=20120616
                      ##source=dbSNP
                      ##dbSNP_BUILD_ID=137
                      ##reference=GRCh37.p5
    path_dbSNP = path_dbSNP5
    # A.iii) Exon Capture File (Intervals list)
      path_exon_capture_file1 = "/home/ueb/Data/BED/TruSeq_exome_targeted_regions.hg19.bed.chr"
      path_exon_capture_file2 = "/home/ueb/Data/BED/all_captured_exomes_from_ucsc_hg19.bed"
      path_exon_capture_file3 = "/home/ueb/Data/Data_Genomes/forGATK/whole_genome_interval_list.hg19.bed"
      path_exon_capture_file4 = "/home/ueb/Data/Data_Genomes/forGATK/whole_genome_interval_list.hg19.bed"
    path_exon_capture_file = path_exon_capture_file4 
  # B. Program file paths
  path_fastq = "/home/ueb/fastqc/fastqc" 
  path_vcfutils = "/usr/share/samtools/vcfutils.pl"
  path_convert2annovar = "/home/ueb/annovar/convert2annovar.pl"
  path_annotate_variation = "/home/ueb/annovar/annotate_variation.pl"
  path_annotate_humandb = "/home/ueb/annovar/humandb/"
  path_summarize_annovar = "/home/ueb/annovar/summarize_annovar.pl"           
  path_snpEff = "/home/ueb/snpEff/" # end with trailing slash but no script, since the same folder is used for several files          
  path_gatk = "/home/ueb/GenomeAnalysisTKLite-2.3-4-gb8f1308/GenomeAnalysisTKLite.jar" # end with the jar file since it can be lite or not
  path_gatk_key = "/home/ueb/GenomeAnalysisTKLite-2.3-4-gb8f1308/ueb_vhir.org.key" # key to avoid gatk 'calling home' (to gatk authors) on each run. See http://gatkforums.broadinstitute.org/discussion/1250/what-is-phone-home-and-how-does-it-affect-me
  path_gatk_sortbyref = "/home/ueb/Data/gatk-data-2.3/SortByRef.pl" # see http://gatkforums.broadinstitute.org/discussion/1328/script-for-sorting-an-input-file-based-on-a-reference-sortbyref-pl
  path_picard = "/home/ueb/software/picard/picard-tools-1.89/MarkDuplicates.jar"           
      
} else if (p_server==2) { # B52 server
    
  path_fastq = "/home/ueb/software/FastQC/fastqc"
  path_genome = "/home/ueb/Data/Data_Genomes/hg19.fa" 
  #path_genome = "/home/ueb/Data/Data_Genomes/rn4/rn4.fa"
  path_vcfutils = "/usr/share/samtools/vcfutils.pl"
  path_convert2annovar = "/home/ueb/software/annovar/convert2annovar.pl"
  path_annotate_variation = "/home/ueb/software/annovar/annotate_variation.pl"
  path_annotate_humandb = "/home/ueb/software/annovar/humandb/"
  path_summarize_annovar = "/home/ueb/software/annovar/summarize_annovar.pl"           
  path_snpEff = "/home/ueb/snpEff/" # end with trailing slash but no script, since the same folder is used for several files           
  path_gatk = "/home/ueb/GenomeAnalysisTKLite-2.3-4-gb8f1308/GenomeAnalysisTKLite.jar" # end with the jar file since it can be lite or not
  path_gatk_key = "/home/ueb/GenomeAnalysisTKLite-2.3-4-gb8f1308/ueb_vhir.org.key" # key to avoid gatk 'calling home' (to gatk authors) on each run. See http://gatkforums.broadinstitute.org/discussion/1250/what-is-phone-home-and-how-does-it-affect-me
  path_picard = "/home/ueb/software/picard/picard-tools-1.89/MarkDuplicates.jar"           
  path_dbSNP1 = "/home/ueb/Data/dbSNP/dbsnp132_20101103.vcf"
  path_dbSNP = path_dbSNP1
  path_exon_capture_file1 = "/home/ueb/Data/BED/TruSeq_exome_targeted_regions.hg19.bed.chr"
  path_exon_capture_file2 = "/home/ueb/Data/BED/all_captured_exomes_from_ucsc_hg19.bed"  
  path_exon_capture_file = path_exon_capture_file2
}
